src/meta/src/stream/stream_graph/fragment.rs

```diff
@@ -522,85 +533,145 @@ impl CompleteStreamFragmentGraph {
                 original_table_fragment_id,
                 downstream_fragments,
             }),
+            None,
         )
     }
 
     fn build_helper(
-        graph: StreamFragmentGraph,
+        mut graph: StreamFragmentGraph,
         upstream_ctx: Option<FragmentGraphUpstreamContext>,
         downstream_ctx: Option<FragmentGraphDownstreamContext>,
+        table_job_type: Option<TableJobType>,
     ) -> MetaResult<Self> {
         let mut extra_downstreams = HashMap::new();
         let mut extra_upstreams = HashMap::new();
         let mut existing_fragments = HashMap::new();
 
         if let Some(FragmentGraphUpstreamContext {
-            upstream_mview_fragments,
+            upstream_job_fragments,
         }) = upstream_ctx
         {
             // Build the extra edges between the upstream `Materialize` and the downstream `Chain`
             // of the new materialized view.
-            for (&id, fragment) in &graph.fragments {
+            for (&id, fragment) in &mut graph.fragments {
                 for (&upstream_table_id, output_columns) in &fragment.upstream_table_columns {
-                    let mview_fragment = upstream_mview_fragments
-                        .get(&upstream_table_id)
-                        .context("upstream materialized view fragment not found")?;
-                    let mview_id = GlobalFragmentId::new(mview_fragment.fragment_id);
-
-                    // Resolve the required output columns from the upstream materialized view.
-                    let output_indices = {
-                        let nodes = mview_fragment.actors[0].get_nodes().unwrap();
-                        let mview_node = nodes.get_node_body().unwrap().as_materialize().unwrap();
-                        let all_column_ids = mview_node
-                            .get_table()
-                            .unwrap()
-                            .columns
-                            .iter()
-                            .map(|c| c.column_desc.as_ref().unwrap().column_id)
-                            .collect_vec();
-
-                        output_columns
-                            .iter()
-                            .map(|c| {
-                                all_column_ids
+                    let (up_fragment_id, edge) = match table_job_type.as_ref() {
+                        Some(TableJobType::SharedCdcSource) => {
```

Comment:
I still don't quite like introducing a new `DispatcherType` for such cases. 😕 From my perspective, it breaks the readability and maintainability a lot.